addons/binding/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/protocol/RemoteControllerWebSocket.java

```diff
@@ -0,0 +1,678 @@
+                        if (currentSourceApp == null || currentSourceApp.isEmpty()) {
+                            callback.currentAppUpdated("");
+                        }
+                    }
+                } else if (json.getAsString("event") != null) {
+                    switch (json.getAsString("event")) {
+                        case "ms.channel.connect":
+                            logger.debug("Remote channel connected");
+                            // update is requested from ed.installedApp.get event: small risk that this websocket is not
+                            // yet connected
+                            break;
+                        case "ms.channel.clientConnect":
+                            logger.debug("Remote client connected");
+                            break;
+                        case "ms.channel.clientDisconnect":
+                            logger.debug("Remote client disconnected");
+                            break;
+                        default:
+                            logger.debug("WebSocketRemote Unknown event: {}", msg);
+
+                    }
+                }
+            } catch (Exception e) {
+                logger.error("{}: Error ({}) in message: {}", this.getClass().getSimpleName(), e.getMessage(), msg, e);
+            }
+        }
+
+    }
+
+    /**
+     * Websocket class to retrieve artmode status (on o.a. the Frame TV's)
+     *
+     * @author Arjan Mels
+     *
+     */
+    class WebSocketArt extends WebSocketBase {
+        @Override
+        public void onWebSocketText(String msgarg) {
+            String msg = msgarg.replace('\n', ' ');
+            super.onWebSocketText(msg);
+            try {
+                JSONObject json = (JSONObject) new JSONParser(JSONParser.MODE_PERMISSIVE).parse(msg);
+                switch (json.getAsString("event")) {
+                    case "ms.channel.connect":
+                        logger.debug("Art channel connected");
+                        break;
+                    case "ms.channel.ready":
+                        logger.debug("Art channel ready");
+                        getArtmodeStatus();
+                        break;
+                    case "ms.channel.clientConnect":
+                        logger.debug("Art client connected");
+                        break;
+                    case "ms.channel.clientDisconnect":
+                        logger.debug("Art client disconnected");
+                        break;
+
+                    case "d2d_service_message":
+                        JSONObject json2 = (JSONObject) new JSONParser(JSONParser.MODE_PERMISSIVE)
+                                .parse(json.getAsString("data"));
+                        if (json2 == null || json2.getAsString("event") == null) {
+                            logger.debug("Empty d2d_service_message event: {}", msg);
+                        } else {
+                            switch (json2.getAsString("event")) {

```

Comment:
Should we nest switches or can this be its own method?

Reply:
done